Unstable test executor/asyncloaddata TestAsyncLoad

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->

```
make failpoint-enable
cd executor/asyncloaddata
go test -tags intest -run TestAsyncLoad
``` 

### 2. What did you expect to see? (Required)

No error

### 3. What did you see instead (Required)

```
--- FAIL: TestAsyncLoad (27.91s)
    --- FAIL: TestAsyncLoad/TestInternalStatus (14.07s)
        show_test.go:284:
                Error Trace:    /home/genius/project/src/github.com/pingcap/tidb/executor/asyncloaddata/show_test.go:284
                                                        /home/genius/project/go/src/runtime/asm_amd64.s:1598
                Error:          "[]" should have 1 item(s), but has 0
                Test:           TestAsyncLoad/TestInternalStatus
FAIL
exit status 1
FAIL    github.com/pingcap/tidb/executor/asyncloaddata  27.967s
```

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->

master 6273e22b04cc61c4a1ec4640f97c7810fa36c97a

--- a/executor/asyncloaddata/BUILD.bazel
+++ b/executor/asyncloaddata/BUILD.bazel
@@ -38,6 +38,7 @@ go_test(
     shard_count = 6,
     deps = [
         "//br/pkg/lightning/config",
+        "//executor",
         "//executor/importer",
         "//kv",
         "//parser/auth",

--- a/executor/asyncloaddata/show_test.go
+++ b/executor/asyncloaddata/show_test.go
@@ -25,6 +25,7 @@ import (
 	"github.com/fsouza/fake-gcs-server/fakestorage"
 	"github.com/pingcap/failpoint"
 	"github.com/pingcap/tidb/br/pkg/lightning/config"
+	"github.com/pingcap/tidb/executor"
 	. "github.com/pingcap/tidb/executor/asyncloaddata"
 	"github.com/pingcap/tidb/executor/importer"
 	"github.com/pingcap/tidb/kv"
@@ -269,15 +270,14 @@ func (s *mockGCSSuite) TestInternalStatus() {
 	wg.Add(1)
 	go func() {
 		defer wg.Done()
+		defer close(executor.TestSyncCh)
+
 		tk2 := testkit.NewTestKit(s.T(), s.store)
 		tk2.Session().GetSessionVars().User = user
 		userStr := tk2.Session().GetSessionVars().User.String()
 
-		// tk @ 0:00
-		// create load data job record in the system table and sleep 3 seconds
-
-		time.Sleep(2 * time.Second)
-		// tk2 @ 0:02
+		// wait for the load data job to be created
+		<-executor.TestSyncCh
 
 		jobInfos, err := GetAllJobInfo(ctx, tk2.Session(), userStr)
 		require.NoError(s.T(), err)
@@ -318,11 +318,11 @@ func (s *mockGCSSuite) TestInternalStatus() {
 		}
 		r.checkIgnoreTimes(s.T(), row)
 
-		// tk @ 0:03
-		// start job and sleep 3 seconds
+		// resume the load data job
+		executor.TestSyncCh <- struct{}{}
 
-		time.Sleep(3 * time.Second)
-		// tk2 @ 0:05
+		// wait for the load data job to be started
+		<-executor.TestSyncCh
 
 		info, err = GetJobInfo(ctx, tk2.Session(), id, userStr)
 		require.NoError(s.T(), err)
@@ -336,12 +336,20 @@ func (s *mockGCSSuite) TestInternalStatus() {
 		r.jobStatus = "running"
 		r.checkIgnoreTimes(s.T(), row)
 
-		// tk @ 0:06
-		// commit one task and sleep 3 seconds
+		// resume the load data job
+		executor.TestSyncCh <- struct{}{}
 
-		time.Sleep(3 * time.Second)
-		// tk2 @ 0:08
+		// wait for the first task to be committed
+		<-executor.TestSyncCh
 
+		// wait for UpdateJobProgress
+		require.Eventually(s.T(), func() bool {
+			info, err = GetJobInfo(ctx, tk2.Session(), id, userStr)
+			if err != nil {
+				return false
+			}
+			return info.Progress == `{"SourceFileSize":2,"LoadedFileSize":1,"LoadedRowCnt":1}`
+		}, 6*time.Second, time.Millisecond*100)
 		info, err = GetJobInfo(ctx, tk2.Session(), id, userStr)
 		require.NoError(s.T(), err)
 		expected.Progress = `{"SourceFileSize":2,"LoadedFileSize":1,"LoadedRowCnt":1}`
@@ -354,16 +362,20 @@ func (s *mockGCSSuite) TestInternalStatus() {
 		r.loadedFileSize = "1B"
 		r.checkIgnoreTimes(s.T(), row)
 
-		// tk @ 0:09
-		// commit one task and sleep 3 seconds
+		// resume the load data job
+		executor.TestSyncCh <- struct{}{}
 
-		time.Sleep(3 * time.Second)
-		// tk2 @ 0:11
+		// wait for the second task to be committed
+		<-executor.TestSyncCh
 
-		info, err = GetJobInfo(ctx, tk2.Session(), id, userStr)
-		require.NoError(s.T(), err)
-		expected.Progress = `{"SourceFileSize":2,"LoadedFileSize":2,"LoadedRowCnt":2}`
-		require.Equal(s.T(), expected, info)
+		// wait for UpdateJobProgress
+		require.Eventually(s.T(), func() bool {
+			info, err = GetJobInfo(ctx, tk2.Session(), id, userStr)
+			if err != nil {
+				return false
+			}
+			return info.Progress == `{"SourceFileSize":2,"LoadedFileSize":2,"LoadedRowCnt":2}`
+		}, 6*time.Second, time.Millisecond*100)
 
 		rows = tk2.MustQuery(fmt.Sprintf("SHOW LOAD DATA JOB %d;", id)).Rows()
 		require.Len(s.T(), rows, 1)
@@ -371,11 +383,16 @@ func (s *mockGCSSuite) TestInternalStatus() {
 		r.loadedFileSize = "2B"
 		r.checkIgnoreTimes(s.T(), row)
 
-		// tk @ 0:12
-		// finish job
+		// resume the load data job
+		executor.TestSyncCh <- struct{}{}
 
-		time.Sleep(3 * time.Second)
-		// tk2 @ 0:14
+		require.Eventually(s.T(), func() bool {
+			info, err = GetJobInfo(ctx, tk2.Session(), id, userStr)
+			if err != nil {
+				return false
+			}
+			return info.Status == JobFinished
+		}, 6*time.Second, 100*time.Millisecond)
 
 		info, err = GetJobInfo(ctx, tk2.Session(), id, userStr)
 		require.NoError(s.T(), err)
@@ -410,9 +427,9 @@ func (s *mockGCSSuite) TestInternalStatus() {
 		config.BufferSizeScale = backup3
 	})
 
-	s.enableFailpoint("github.com/pingcap/tidb/executor/AfterCreateLoadDataJob", `sleep(3000)`)
-	s.enableFailpoint("github.com/pingcap/tidb/executor/AfterStartJob", `sleep(3000)`)
-	s.enableFailpoint("github.com/pingcap/tidb/executor/AfterCommitOneTask", `sleep(3000)`)
+	s.enableFailpoint("github.com/pingcap/tidb/executor/SyncAfterCreateLoadDataJob", `return`)
+	s.enableFailpoint("github.com/pingcap/tidb/executor/SyncAfterStartJob", `return`)
+	s.enableFailpoint("github.com/pingcap/tidb/executor/SyncAfterCommitOneTask", `return`)
 	sql := fmt.Sprintf(`LOAD DATA INFILE 'gs://test-tsv/t*.tsv?endpoint=%s'
 		INTO TABLE load_tsv.t WITH batch_size = 1;`, gcsEndpoint)
 	s.tk.MustExec(sql)

--- a/executor/load_data.go
+++ b/executor/load_data.go
@@ -310,6 +310,9 @@ func (e *LoadDataWorker) Load(
 	return jobID, e.doLoad(ctx, readerInfos, jobID)
 }
 
+// TestSyncCh is used in unit test to synchronize the execution of LOAD DATA.
+var TestSyncCh = make(chan struct{})
+
 func (e *LoadDataWorker) doLoad(
 	ctx context.Context,
 	readerInfos []importer.LoadDataReaderInfo,
@@ -359,6 +362,10 @@ func (e *LoadDataWorker) doLoad(
 	}()
 
 	failpoint.Inject("AfterCreateLoadDataJob", nil)
+	failpoint.Inject("SyncAfterCreateLoadDataJob", func() {
+		TestSyncCh <- struct{}{}
+		<-TestSyncCh
+	})
 
 	totalFilesize := int64(0)
 	hasErr := false
@@ -380,6 +387,10 @@ func (e *LoadDataWorker) doLoad(
 	}
 
 	failpoint.Inject("AfterStartJob", nil)
+	failpoint.Inject("SyncAfterStartJob", func() {
+		TestSyncCh <- struct{}{}
+		<-TestSyncCh
+	})
 
 	group, groupCtx := errgroup.WithContext(ctx)
 	// done is used to let commitWork goroutine notify UpdateJobProgress
@@ -698,6 +709,10 @@ func (w *commitWorker) commitWork(ctx context.Context, inCh <-chan commitTask) (
 				zap.Uint64("taskCnt processed", taskCnt),
 			)
 			failpoint.Inject("AfterCommitOneTask", nil)
+			failpoint.Inject("SyncAfterCommitOneTask", func() {
+				TestSyncCh <- struct{}{}
+				<-TestSyncCh
+			})
 		}
 	}
 }

